### PR TITLE
[RDY] Add complete() noinsert/noselect support

### PR DESCRIPTION
**Pull request by Shougo**
https://github.com/neovim/neovim/pull/2564 is merged, but `complete()` does not support the feature.
So, I added the support for `complete()`.
And I added the tests for `complete()`.  Please check it.

### CODE REVIEW

**Comment by splinterofchaos** [comment]
LGTM :+1:

**Comment by fmoralesc** [comment]
LGTM too. Thanks for following up on this, @Shougo.

File: test/functional/viml/completion_spec.lua [hunk 1]
```diff
@@ -54,25 +55,42 @@ describe('completion', function()
...
+      eq('foo', eval('getline(3)'))
     end)
     it('does not insert the first candidate if noselect', function()
       execute('set completeopt+=menuone,noselect')
       feed('ifoo<ESC>o<C-x><C-n>bar<ESC>')
       eq('bar', eval('getline(2)'))
+      feed('o<C-r>=TestComplete()<CR>bar<ESC>')
+      eq('bar', eval('getline(3)'))
     end)
     it('does not select/insert the first candidate if noselect and noinsert', function()
```
Comment (oni-link) [inline test/functional/viml/completion_spec.lua]: Wondering why `noselect` and `noinsert` are tested together.
`noselect` is enough to not select and not insert. If `noselect` is used in `completeopt`, `noinsert` has no effect (`compl_no_insert` will always be `FALSE`).

Comment (Shougo) [inline test/functional/viml/completion_spec.lua]: Yes.  But it is implementation feature.
I think it should be tested.

**Comment by Shougo** [comment]
Changed to [RDY].

**Comment by justinmk** [comment]
Build failures were timeouts; restarted.

**Comment by Shougo** [comment]
@justinmk When it is merged?  This feature is needed for async completion plugins.

**Comment by splinterofchaos** [comment]
@Shougo Two LGTMs, and four participants with no complaints, I don't think we need to delay this, but it does need a rebase. That should also fix the timeout on the gcc build thanks to #2806. I'm always tempted to wait another day for comments, but everyone's had a week so I'll merge it tonight if you do this before I go to bed.

File: src/nvim/edit.c [hunk 2]
```diff
@@ -2980,16 +2987,17 @@ static int ins_compl_prep(int c)
...
-      compl_no_select = FALSE;
-    }
+  }
+
+  if (strstr((char *)p_cot, "noselect") != NULL) {
+    compl_no_insert = FALSE;
+    compl_no_select = TRUE;
+  } else if (strstr((char *)p_cot, "noinsert") != NULL) {
+    compl_no_insert = TRUE;
+    compl_no_select = FALSE;
```
Comment (splinterofchaos) [inline src/nvim/edit.c]: Are `noinsert` and `noselect` mutually exclusive? I'm wondering why this is `else if` instead of `if`.

Comment (Shougo) [inline src/nvim/edit.c]: It is not exclusive option.
User can set both values.
But `compl_no_insert` and `compl_no_select` are exclusive variables.

So "else if" is needed.

**Comment by Shougo** [comment]
@splinterofchaos I rebased it.

**Comment by splinterofchaos** [comment]
landed in be66c0b3570a55671656959203bd5b824f77fde7

Thanks, and good luck with deocomplete!

**Comment by justinmk** [comment]
Does this need user documentation?

**Comment by Shougo** [comment]
@justinmk I think the documentation is available in https://github.com/neovim/neovim/pull/2564.

**Comment by Shougo** [comment]
> Thanks, and good luck with deocomplete!

Thanks!

**Pull request closed without merging**